File: library/src/main/java/com/crashops/sdk/communication/Reachability.java
package com.crashops.sdk.communication;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.wifi.WifiManager;
import android.os.Build;
import androidx.annotation.Nullable;

import com.crashops.sdk.CrashOps;
import com.crashops.sdk.util.PrivateEventBus;

public enum Reachability {
    instance;

    Reachability() {
        wifiManager = (WifiManager) CrashOps.getApplicationContext().getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        connectivityManager = (ConnectivityManager) CrashOps.getApplicationContext().getSystemService(Context.CONNECTIVITY_SERVICE);
        if (connectivityManager != null) {
            NetworkInfo wifiNetworkInfo = connectivityManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
        }

    }

    public enum NetworkStatus {
        Unknown, Cellular, WiFi, None;

        @Override
        public String toString() {
            return name();
        }
    }

    private NetworkStatus status;

    @Nullable
    private final ConnectivityManager connectivityManager;
    @Nullable
    private final WifiManager wifiManager;

    public NetworkStatus getReachabilityStatus() {
        status = NetworkStatus.Unknown;
        Boolean isConnectedToWifi = isConnectedToWifi();
        Boolean isConnectedToCellularData = isConnectedToCellularData();
        if (isConnectedToWifi != null && isConnectedToWifi) {
            status = NetworkStatus.WiFi;
        } else if (isConnectedToCellularData != null && isConnectedToCellularData) {
            status = NetworkStatus.Cellular;
        } else if (wifiManager != null && connectivityManager != null) {
            status = NetworkStatus.None;
        }

        return status;
    }

    public NetworkStatus getStatus() {
        return status;
    }

    @Nullable
    Boolean isConnectedToCellularData() {
        if (connectivityManager == null) return null;
        NetworkInfo ni = connectivityManager.getNetworkInfo(ConnectivityManager.TYPE_MOBILE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            connectivityManager.addDefaultNetworkActiveListener(new ConnectivityManager.OnNetworkActiveListener() {
                @Override
                public void onNetworkActive() {
                    PrivateEventBus.notify(PrivateEventBus.Action.ON_NETWORK_ACTIVE);
                }
            });
        }

        // The 'ni == null' means that this device does not have mobile data capability
        return ni != null && ni.isAvailable();
    }

    @Nullable
    private Boolean isConnectedToWifi() {
        if (connectivityManager == null) return null;
        NetworkInfo ni = connectivityManager.getNetworkInfo(ConnectivityManager.TYPE_WIFI);

        // The 'ni == null' means that this device does not have mobile data capability
        return ni != null && ni.isAvailable();
    }

    @Nullable
    private Boolean isWifiEnabled() {
        if (wifiManager == null) return null;
        return wifiManager.isWifiEnabled();
    }
}

File: library/src/main/java/com/crashops/sdk/communication/Reachability.kt
package com.crashops.sdk.communication

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build

class Reachability {
    enum class NetworkStatus {
        Unknown, Cable, Cellular, WiFi, None;

        override fun toString(): String {
            return name
        }
    }

    companion object {
        @JvmStatic
        public fun getReachabilityStatus(context: Context): NetworkStatus {
            var result = NetworkStatus.Unknown
            val connectivityManager =
                    context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                val networkCapabilities = connectivityManager.activeNetwork ?: return result
                val actNw =
                        connectivityManager.getNetworkCapabilities(networkCapabilities) ?: return result
                result = when {
                    actNw.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> NetworkStatus.WiFi
                    actNw.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> NetworkStatus.Cellular
                    actNw.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> NetworkStatus.Cable
                    else -> NetworkStatus.None
                }
            } else {
                connectivityManager.run {
                    connectivityManager.activeNetworkInfo?.run {
                        result = when (type) {
                            ConnectivityManager.TYPE_WIFI -> NetworkStatus.WiFi
                            ConnectivityManager.TYPE_MOBILE -> NetworkStatus.Cellular
                            ConnectivityManager.TYPE_ETHERNET -> NetworkStatus.Cable
                            else -> NetworkStatus.None
                        }

                    }
                }
            }

            return result
        }
    }
}